crates/webidl/src/lib.rs

```diff
@@ -65,23 +66,31 @@ fn compile_ast(ast: &backend::ast::Program) -> String {
     tokens.to_string()
 }
 
-trait WebidlParse {
-    fn webidl_parse(&self, program: &mut backend::ast::Program) -> Result<()>;
+trait WebidlParse<'a> {
+    type Extra;
+
+    fn webidl_parse(&self, program: &mut backend::ast::Program, extra: Self::Extra) -> Result<()>;
 }
 
-impl WebidlParse for Vec<webidl::ast::Definition> {
-    fn webidl_parse(&self, program: &mut backend::ast::Program) -> Result<()> {
+impl<'a> WebidlParse<'a> for Vec<webidl::ast::Definition> {
+    type Extra = ();
+
+    fn webidl_parse(&self, program: &mut backend::ast::Program, _: ()) -> Result<()> {
         for def in self {
-            def.webidl_parse(program)?;
+            def.webidl_parse(program, ())?;
         }
         Ok(())
     }
 }
 
-impl WebidlParse for webidl::ast::Definition {
-    fn webidl_parse(&self, program: &mut backend::ast::Program) -> Result<()> {
+impl<'a> WebidlParse<'a> for webidl::ast::Definition {
+    type Extra = ();
```

Comment:
Hm once you got down the route of associated types here and `()` all over the place it may just be best to have a `Context` struct with shared state and a bunch of methods for each recursive step. Either way's fine by me, but I tend to shy away from traits too aggressively historically

Reply:
I like `Context` for shared state that is uniform across all `impl`s and `Self::Extra` for things that change between different `impl`s.

I find that traits help me clarify my thinking about what is going on, even if they end up varying a bit because of the `Extra` type.